### DISCUSSION TITLE
PlatformNotSupported Exception occurs using Microsoft.Extensions.DependencyInjection on WASM (net6/WinUI)

### DISCUSSION THREAD

**Discussion by ToddGlodek** [unoplatform/uno#9961] [Q&A]
There is documentation on the [Uno Platform website indicating that Microsoft.Extensions.DependencyInjection](https://www.andrewhoefling.com/Blog/Post/mvvm-and-dependency-injection-in-uno-platform-ios-android-wasm-uwp) can be used in projects.   That discussion refers to an [Andrew Hoefling article regarding MVVM and Dependency Injection in Uno Platform](https://www.andrewhoefling.com/Blog/Post/mvvm-and-dependency-injection-in-uno-platform-ios-android-wasm-uwp).  The Hoefling  article is from Oct. 2020 - and specifically covers using Microsoft.Extensions.DependencyInjection on WASM.  However, the [code samples provided relate](https://github.com/SkyeHoefling/Uno-Platform-Samples/tree/main/MvvmAndDependencyInjection) to UWP and use :    `<TargetFramework>netstandard2.0</TargetFramework>`  

When using     `<TargetFramework>net6.0</TargetFramework>`   I encounter a PlatformNotSupported exception at runtime when calling `DependenciesContainer.Current.GetService<MySampleClass>(); ` on WASM.  However, the code runs just fine on the WinUI project.

Specifically the error is :    System.PlatformNotSupportedException: **Cannot wait on monitors on this runtime**.

>  at System.Threading.Monitor.ObjWait(Int32 millisecondsTimeout, Object obj) in D:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\mono\System.Private.CoreLib\src\System\Threading\Monitor.Mono.cs:line 121
> 
> ... 
>  Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2[[Microsoft.Extensions.DependencyInjection.ServiceLookup.RuntimeResolverContext, Microsoft.Extensions.DependencyInjection, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60],[System.Object, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].VisitCallSiteMain(ServiceCallSite callSite, RuntimeResolverContext argument)
  
I have applied all the latest Nuget packages (including pre-release) - but that doesn't have any effect:
```
>     <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="6.0.2-mauipre.1.22102.15" />
>     <PackageReference Include="Microsoft.Extensions.Logging" Version="6.0.2-mauipre.1.22102.15" />
>     <PackageReference Include="Microsoft.Windows.Compatibility" Version="6.0.2-mauipre.1.22102.15" />
>     <PackageReference Include="Newtonsoft.Json" Version="13.0.2-beta2" />
>     <PackageReference Include="Uno.Extensions.Logging.WebAssembly.Console" Version="1.4.0" />
>     <PackageReference Include="Uno.WinUI.WebAssembly" Version="4.5.9" />
>     <PackageReference Include="Uno.WinUI.RemoteControl" Version="4.5.9" Condition="'$(Configuration)'=='Debug'" />
>     <PackageReference Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="4.5.9" />
>     <PackageReference Include="Uno.Wasm.Bootstrap" Version="3.3.1" />
>     <PackageReference Include="Uno.Wasm.Bootstrap.DevServer" Version="3.3.1" 
```

If someone out there has gotten this to work with `<TargetFramework>net6.0</TargetFramework>` can you please point me to some documentation / repositories I can look at.  I know that Microsoft.Extensions.DependencyInjection in not a Uno library, so I didn't open this discussion as a BUG/ISSUE.   However, since it is one of the most popular libraries for cross platform development - any special work arounds that needs to be done before it be used **should** be thoroughly covered in the Uno documentation.

UPDATE :  I made the changes that were described in [Uno Platform 4.4 – Wasm Threading+Exception Handling](https://platform.uno/blog/uno-platform-4-4-wasm-threadingexception-handling-rich-animations-gamepad-apis-and-more/)
Updated these NuGets:
```
    <PackageReference Include="Uno.Wasm.Bootstrap" Version="4.0.0-dev.193" />
    <PackageReference Include="Uno.Wasm.Bootstrap.DevServer" Version="4.0.0-dev.193" />
```
Modified the WASM.csproj file setting this value to TRUE 
`   <WasmShellILLinkerEnabled>true</WasmShellILLinkerEnabled>`

I think that fixed it !!

**Comment by jeromelaban** [comment]
Thanks for the report. Setting `WasmShellILLinkerEnabled` likely hides an issue because it removes code that may otherwise be invalid. If you have a simple repro you can attach here, someone may be able to take a look.
